### PR TITLE
feat: add C implementation for `math/base/special/rempio2`

### PR DESCRIPTION
**Pull request by gunjjoshi**
## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [`math/base/special/rempio2`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/rempio2).

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #649.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

This is a draft PR, to discuss the implementation.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by gunjjoshi** [comment]
I was getting failure in some of the test cases. I have referred to FreeBSD version `12.2.0`, as a reference. 

`rempio2`: https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_rem_pio2.c?revision=367086&view=markup
`kernel-rempio2`: https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/k_rem_pio2.c?revision=367086&view=markup

Specifically, if we can check `main.c`, `addon.c` and `native.js`, that would help us out in completing this.

File: lib/node_modules/@stdlib/math/base/special/rempio2/manifest.json [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+			],
+			"libraries": [],
+			"libpath": [],
+			"dependencies": [
+				"@stdlib/math/base/special/round",
+				"@stdlib/math/base/special/floor",
+				"@stdlib/math/base/special/ldexp",
+				"@stdlib/number/float64/base/get_high_word",
+				"@stdlib/number/float64/base/get_low_word",
+				"@stdlib/number/float64/base/from_words"
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/rempio2/manifest.json]:
```suggestion
				"@stdlib/number/float64/base/get-high-word",
				"@stdlib/number/float64/base/get-low-word",
				"@stdlib/number/float64/base/from-words"
```

File: lib/node_modules/@stdlib/math/base/special/rempio2/manifest.json [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/rempio2/manifest.json]:
```suggestion
				"@stdlib/number/float64/base/get-high-word",
				"@stdlib/number/float64/base/get-low-word",
				"@stdlib/number/float64/base/from-words"
```

**Comment by kgryte** [comment]
@gunjjoshi I refactored the addon (see https://github.com/stdlib-js/stdlib/pull/2279/commits/0df78ac6dd54e79bdb066f7196fe127a92e0ec29). Tests now pass; however, results for C and JS differ. Currently, if provided `128.0`, JS returns `~3.618e-17` and C returns `-1.6380356615247201e-19`. You'll want to investigate why this is the case.

**Comment by gunjjoshi** [comment]
> @gunjjoshi I refactored the addon (see [0df78ac](https://github.com/stdlib-js/stdlib/commit/0df78ac6dd54e79bdb066f7196fe127a92e0ec29)). Tests now pass; however, results for C and JS differ. Currently, if provided `128.0`, JS returns `~3.618e-17` and C returns `-1.6380356615247201e-19`. You'll want to investigate why this is the case.

Great, thanks for the help! I'll go through the implementation once more to check everything.

**Comment by gunjjoshi** [comment]
@kgryte I was able to track down the problem. This was because in the `C` implementation, earlier the exponent mask the we used had a value of `0x7ff00000`. But in the `javascript` implementation, we were using `0x7ff`.

After the changes, both the implementations give the same result for `128.0`, i.e., `~3.618e-17`.

Also, the lint error that we currently have is arising due to two similar `if` conditions, but that is required for our implementation.

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+		// Check if a second iteration is needed (good to 118 bits)...
+		if ( i > 16 ) {
+			t = r;
+			w = n * PIO2_2;
+			r = t - w;
+			w = ( n * PIO2_2T ) - ( ( t - r ) - w );
+			*rem1 = r - w;
+			stdlib_base_float64_get_high_word( *rem1, &uhigh );
+			high = ( int32_t )uhigh;
+			i = j - ( ( high >> 20 ) & STDLIB_CONSTANT_FLOAT64_HIGH_WORD_EXPONENT_MASK );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c]:
```suggestion
			i = j - ( ( high >> 20 ) & EXPONENT_MASK );
```

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+				r = t - w;
+				w = ( n * PIO2_3T ) - ( ( t - r ) - w );
+				*rem1 = r - w;
+			}
+		}
+		*rem2 = ( r - *rem1 ) - w;
+		return n;
+	}
+	// Case: x is NaN or infinity
+	if( ix >= STDLIB_CONSTANT_FLOAT64_HIGH_WORD_EXPONENT_MASK ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c:599]: This one is correct.

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c (hunk 3 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c]: Notice here that we are also shifting `high` to the right by 20 bits, thus bringing the exponent section of the high word to the final bits. In which case, we need to use `0x7ff`, not the predefined constant.

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+	2.16741683877804819444e-51  // 0x3569F31D, 0x00000000
+};
+
+// Arrays for storing temporary values:
+double TX[ 3 ] = { 0.0 };
+double TY[ 2 ] = { 0.0 };
+double F[ 20 ];
+double Q[ 20 ];
+double FQ[ 20 ];
+int32_t IQ[ 20 ];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c]: These need to be moved to within the implementation. In C, we need to worry about thread safety. See, e.g., https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/k_rem_pio2.c?view=markup#l297.

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+
+/**
+* Returns the last three binary digits of `N` with `y = x - Nπ/2` so that `|y| < π/2`.
+*
+* ## Method
+*
+* -   The method is to compute the integer (`mod 8`) and fraction parts of `2x/π` without doing the full multiplication. In general, we skip the part of the product that is known to be a huge integer (more accurately, equals `0 mod 8` ). Thus, the number of operations is independent of the exponent of the input.
+*
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c:228]: We're also missing the extended comment in the source code. This is likely also missing in the JS implementation and can be added in this PR. https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/k_rem_pio2.c?view=markup#l33

Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c:228]: I've added the comment in both files. Is the format fine?

**Comment by gunjjoshi** [comment]
There is a lint error saying the arrays are uninitialized. So, should we initialize them with zeros at the time of declaration directly, or should we declare them first, then initialize them with zeros using a `for` loop later?

**Comment by kgryte** [comment]
@gunjjoshi I'd go ahead and initialize with zeros at the time of declaration.

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c (hunk 6 above)
Comment (Planeshifter) [reply lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c:228]: @gunjjoshi Made some small edits and nested unordered lists to satisfy our JSDoc linting rules, which don't like arbitrary indentation.

**Review by Planeshifter — APPROVED** [review]
Let's get this in to unblock additional C implementation work. Will merge once CI has passed.

Thanks, @gunjjoshi, for your work on it!

File: lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c (hunk 6 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/rempio2/src/main.c:228]: Thanks, @Planeshifter, I'll keep this in mind.

**Comment by gunjjoshi** [comment]
Thanks @Planeshifter @kgryte for helping in this.

**Pull request merged**